api/types/container/network.go

```diff
@@ -0,0 +1,243 @@
+package container
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"unique"
+)
+
+type PortProto string
+
+const (
+	TCP  PortProto = "tcp"
+	UDP  PortProto = "udp"
+	SCTP PortProto = "sctp"
+)
+
+// Port is a type representing a single port number and protocol in the format "80/tcp".
+//
+// The zero port value, i.e. Port{}, is invalid; use [ParsePort] to create a valid Port value.
+type Port struct {
+	num   uint16
+	proto PortProto
+}
+
+// zeroPort and zeroPortRange are sentinel [Port] and [PortRange] values.
+var (
+	zeroPort      = unique.Make(Port{})
+	zeroPortRange = unique.Make(PortRange{})
+)
+
+// ParsePort parses s as a [Port].
+func ParsePort(s string) (Port, error) {
+	if s == "" {
+		return Port{}, errors.New("value is empty")
+	}
+
+	port, proto, ok := strings.Cut(s, "/")
+	if !ok {
+		proto = "tcp"
+	}
+
+	portVal, err := parsePortNumber(port)
+	if err != nil {
+		return Port{}, fmt.Errorf("invalid port '%s': %w", port, err)
+	}
+
+	return Port{num: portVal, proto: PortProto(proto)}, nil
+}
+
+// MustParsePort calls [ParsePort](s) and panics on error.
+func MustParsePort(s string) Port {
+	p, err := ParsePort(s)
+	if err != nil {
+		panic(err)
+	}
+	return p
+}
+
+// PortFrom returns a Port with the given number and protocol.
+func PortFrom(num uint16, proto PortProto) Port {
+	return Port{num: num, proto: proto}
+}
+
+func (p Port) Num() uint16 {
+	return p.num
+}
+
+func (p Port) Proto() PortProto {
+	return p.proto
+}
+
+// IsValid reports whether the [Port] is an initialized valid port (not the zero value).
+func (p Port) IsValid() bool {
+	return p != zeroPort.Value()
+}
+
+func (p Port) String() string {
+	switch p {
+	case zeroPort.Value():
+		return "invalid port"
+	default:
+		return fmt.Sprintf("%d/%s", p.num, p.proto)
+	}
+}
+
+func (p Port) MarshalText() ([]byte, error) {
+	return []byte(p.String()), nil
+}
```

Comment:
`Port{}.MarshalText() == []byte("invalid port")` which will not unmarshal successfully. The zero value needs to marshal to a nil or zero-length byte slice.